File: 2025/06 June/14 Symmetric Tree/main.py
# Link: https://www.geeksforgeeks.org/problems/symmetric-tree/1


class Node:
    def __init__(self, val):
        self.right = None
        self.data = val
        self.left = None


def createBinaryTree(s):
    if not s:
        return None

    nodes = s.split()
    if nodes[0] == "N":
        return None

    root = Node(int(nodes[0]))
    queue = [root]
    i = 1

    while i < len(nodes):
        current = queue.pop(0)

        if nodes[i] != "N":
            current.left = Node(int(nodes[i]))
            queue.append(current.left)
        i += 1

        if i < len(nodes) and nodes[i] != "N":
            current.right = Node(int(nodes[i]))
            queue.append(current.right)
        i += 1

    return root


class Solution:
    def isSymmetric(self, root):
        if not root:
            return True

        stack1 = [root.left]
        stack2 = [root.right]

        while stack1 and stack2:
            node1 = stack1.pop()
            node2 = stack2.pop()

            if not node1 and not node2:
                continue

            if not node1 or not node2 or node1.data != node2.data:
                return False

            stack1.append(node1.left)
            stack2.append(node2.right)

            stack1.append(node1.right)
            stack2.append(node2.left)

        return not stack1 and not stack2


# Test cases
if __name__ == "__main__":
    solution = Solution()
    s = "1 2 2 3 4 4 3"
    root = createBinaryTree(s)
    print(solution.isSymmetric(root))

    s = "1 2 2 N 3 N 3"
    root = createBinaryTree(s)
    print(solution.isSymmetric(root))

File: 2025/06 June/14 Symmetric Tree/main.cpp
// Link: https://www.geeksforgeeks.org/problems/symmetric-tree/1

#include <iostream>
#include <vector>
#include <stack>
using namespace std;

class Node
{
public:
    int data;
    Node *left, *right;

    Node(int val)
    {
        data = val;
        left = right = nullptr;
    }
};

Node *createBinaryTree(const string &s)
{
    if (s.empty())
        return nullptr;

    vector<string> nodes;
    size_t pos = 0, found;
    while ((found = s.find(' ', pos)) != string::npos)
    {
        nodes.push_back(s.substr(pos, found - pos));
        pos = found + 1;
    }
    nodes.push_back(s.substr(pos));

    if (nodes[0] == "N")
        return nullptr;

    Node *root = new Node(stoi(nodes[0]));
    vector<Node *> queue = {root};
    int i = 1;

    while (i < nodes.size())
    {
        Node *current = queue.front();
        queue.erase(queue.begin());

        if (nodes[i] != "N")
        {
            current->left = new Node(stoi(nodes[i]));
            queue.push_back(current->left);
        }
        i++;

        if (i < nodes.size() && nodes[i] != "N")
        {
            current->right = new Node(stoi(nodes[i]));
            queue.push_back(current->right);
        }
        i++;
    }

    return root;
}

class Solution
{
public:
    bool isSymmetric(Node *root)
    {
        if (root == nullptr)
            return true;

        stack<Node *> s1, s2;

        s1.push(root->left);
        s2.push(root->right);

        while (!s1.empty() && !s2.empty())
        {
            Node *node1 = s1.top();
            Node *node2 = s2.top();
            s1.pop();
            s2.pop();

            if (node1 == nullptr && node2 == nullptr)
            {
                continue;
            }

            if (node1 == nullptr || node2 == nullptr || node1->data != node2->data)
            {
                return false;
            }

            s1.push(node1->left);
            s2.push(node2->right);

            s1.push(node1->right);
            s2.push(node2->left);
        }

        return s1.empty() && s2.empty();
    }
};

int main()
{
    Solution solution;
    string s = "1 2 2 3 4 4 3";
    Node *root = createBinaryTree(s);
    bool result = solution.isSymmetric(root);
    cout << boolalpha << result << endl;

    s = "1 2 2 N 3 N 3";
    root = createBinaryTree(s);
    result = solution.isSymmetric(root);
    cout << boolalpha << result << endl;

    return 0;
}